### PR TITLE
add active_tasks for view builds using version stamps

### PR DESCRIPTION
**Pull request by tonysun83**
## Overview

Active Tasks requires TotalChanges and ChangesDone to show the progress
of long running tasks. This requires `count_changes_since/2` to be
implemented. Unfortunately, that is not easily done  with
foundationdb. This commit replaces TotalChanges with the
versionstamp + the number of docs as a progress indicator. This can
possibly break existing api that relys on TotalChanges. ChangesDone
will still exist, but instead of relying on the current changes seq
it is simply a reflection of how many documents were written by the
updater process.

The new responses look like this:

```
[
    {
        "node": "node1@127.0.0.1",
        "pid": "<0.622.0>",
        "changes_done": 199,
        "current_version_stamp": "8131141649532-0-198",
        "database": "testdb",
        "db_version_stamp": "8131141649532-0-999",
        "design_document": "_design/example",
        "started_on": 1594703583,
        "type": "indexer",
        "updated_on": 1594703586
    }
]

[
    {
        "node": "node1@127.0.0.1",
        "pid": "<0.1030.0>",
        "changes_done": 1000,
        "current_version_stamp": "8131194932916-0-999",
        "database": "testdb",
        "db_version_stamp": "8131194932916-0-999",
        "design_document": "_design/example",
        "started_on": 1594703636,
        "type": "indexer",
        "updated_on": 1594703665
    }
]
```

Note that this only shows one process. IIUC, we can have multiple workers indexing in parallel. I think this approach also takes care of this with the ChangesDone only reflecting how many documents have been indexed for that range.

## Testing recommendations

Tests need to be added. If this approach is accepted, I will add more precise tests. Currently ran manual tests.

## Checklist

- [X] Code is written and works correctly
- [ ] Changes are covered by tests
- [ ] Any new configurable parameters are documented in `rel/overlay/etc/default.ini`
- [ ] A PR for documentation changes has been made in https://github.com/apache/couchdb-documentation

### CODE REVIEW

File: src/couch_views/src/couch_views_indexer.erl [hunk 1]
```diff
@@ -368,6 +382,7 @@ write_docs(TxDb, Mrst, Docs, State) ->
     if LastSeq == false -> ok; true ->
         couch_views_fdb:set_update_seq(TxDb, Sig, LastSeq)
     end,
+    update_tasks(length(Docs), LastSeq),
```
Comment (garrensmith) [inline src/couch_views/src/couch_views_indexer.erl:385]: I think you can use `DocsNumber` since that is the count of the number of docs processed

File: src/couch_views/src/couch_views_indexer.erl [hunk 2]
```diff
@@ -541,3 +556,25 @@ key_size_limit() ->
 
 value_size_limit() ->
     config:get_integer("couch_views", "value_size_limit", ?VALUE_SIZE_LIMIT).
+
+
+update_tasks(NumChanges, false) ->
```
Comment (garrensmith) [inline src/couch_views/src/couch_views_indexer.erl:561]: Can you rename this to `update_active_tasks`

File: src/couch_views/src/couch_views_indexer.erl [hunk 3]
```diff
@@ -231,10 +231,24 @@ maybe_set_build_status(TxDb, Mrst1, _ViewVS, State) ->
 get_update_start_state(TxDb, Mrst, #{db_seq := undefined} = State) ->
     ViewVS = couch_views_fdb:get_creation_vs(TxDb, Mrst),
     ViewSeq = couch_views_fdb:get_update_seq(TxDb, Mrst),
+    DbSeq = fabric2_db:get_update_seq(TxDb),
+    DbVS = fabric2_fdb:seq_to_vs(DbSeq),
```
Comment (garrensmith) [inline src/couch_views/src/couch_views_indexer.erl:235]: Can you move all the related active_tasks logic into its own function. Maybe something like `active_tasks_start`.

File: src/couch_views/src/couch_views_indexer.erl [hunk 4]
```diff
@@ -231,10 +231,24 @@ maybe_set_build_status(TxDb, Mrst1, _ViewVS, State) ->
...
+    ViewVS1 = convert_version_stamp(ViewVS),
+    DbVS1 = convert_version_stamp(DbVS),
+
+    couch_task_status:add_task([
+        {type, indexer},
+        {database, Mrst#mrst.db_name},
+        {design_document, Mrst#mrst.idx_name},
+        {changes_done, 0},
```
Comment (garrensmith) [inline src/couch_views/src/couch_views_indexer.erl:243]: Will changes always be 0? What happens if an index is half-built and the indexer is starting up again?

File: src/couch_views/src/couch_views_indexer.erl [hunk 5]
```diff
@@ -541,3 +556,25 @@ key_size_limit() ->
...
+    VS1 = convert_version_stamp(VS),
+    couch_task_status:update([{changes_done, Changes2},
+        {current_version_stamp, VS1}]).
+
+
+convert_version_stamp({_, Stamp, _, DocNumber}) ->
+    VS = integer_to_list(Stamp) ++ "-" ++ integer_to_list(DocNumber),
+    list_to_binary(VS);
+
+convert_version_stamp(_) ->
```
Comment (garrensmith) [inline src/couch_views/src/couch_views_indexer.erl:579]: What would cause us to reach this state?

**Comment by nickva** [comment]
@tonysun83 

Agree with @garrensmith, for reporting the job on all the nodes, it might be easier to think of something like this:

  * Having an `<<"active_task_status">>` section to each job's data section, already converted to json format.

  * In the http handler we would query all the active jobs and just emit those json objects. All active jobs are conveniently are located under a single `?ACTIVITY` subspace using [get_active_since/3] https://github.com/apache/couchdb/blob/prototype/fdb-layer/src/couch_jobs/src/couch_jobs_fdb.erl#L358) (it was done that way on purpose to make it simple to query all active jobs in order to report _active_tasks easier 😉 ).

 * To get all the possible job types can use [get_types](https://github.com/apache/couchdb/blob/prototype/fdb-layer/src/couch_jobs/src/couch_jobs_fdb.erl#L321). 

 * Since we'd be calling those outside of couch_jobs we could add a new function to `couch_jobs`, something like `get_active_jobs` perhaps.

File: src/couch_views/src/couch_views_indexer.erl (hunk 5 above)
Comment (tonysun83) [reply src/couch_views/src/couch_views_indexer.erl:579]: initially, when we create the versionstamp it's `not_found`:
`ViewVS = couch_views_fdb:get_creation_vs(TxDb, Mrst),`

File: src/couch_views/src/couch_views_indexer.erl (hunk 4 above)
Comment (tonysun83) [reply src/couch_views/src/couch_views_indexer.erl:243]: I didn't have a good solution to this. Mainly that the indexer process has to die first which means the active_task tuple is automatically removed from memory. I was thinking of writing a counter to fdb, but that opened up a new can of worms

Comment (davisp) [reply src/couch_views/src/couch_views_indexer.erl:243]: Historically, `changes_done` refers to how many changes have been processed by this view update. It's for judging how much work is left before the view is up to date. One way to think of this would be if you had a 10,000,000 docs processed for a view, and then went to process another 1,000,000 docs to bring the view up to date, you wouldn't want to start at 99% finished as that'd be misleading to operators.

File: src/couch_views/src/couch_views_indexer.erl [hunk 6]
```diff
@@ -541,3 +556,25 @@ key_size_limit() ->
...
+update_tasks(NumChanges, LastSeq) ->
+    [Changes] = couch_task_status:get([changes_done]),
+    Changes2 = Changes + NumChanges,
+    VS = fabric2_fdb:seq_to_vs(LastSeq),
+    VS1 = convert_version_stamp(VS),
+    couch_task_status:update([{changes_done, Changes2},
+        {current_version_stamp, VS1}]).
+
+
+convert_version_stamp({_, Stamp, _, DocNumber}) ->
```
Comment (davisp) [inline src/couch_views/src/couch_views_indexer.erl:575]: The third tuple element here is the batch id which while generally 0 probably shouldn't be something we should leave out for consistency's sake.

File: src/chttpd/src/chttpd_misc.erl [hunk 7]
```diff
@@ -31,12 +31,14 @@
 
 -include_lib("couch/include/couch_db.hrl").
 -include_lib("couch_mrview/include/couch_mrview.hrl").
+-include_lib("couch_views/include/couch_views.hrl").
```
Comment (tonysun83) [inline src/chttpd/src/chttpd_misc.erl:34]: probably don't need this anymore

File: src/couch_views/src/couch_views_indexer.erl (hunk 2 above)
Comment (tonysun83) [reply src/couch_views/src/couch_views_indexer.erl]: removed, no longer needed

File: src/chttpd/src/chttpd_misc.erl [hunk 8]
```diff
@@ -294,11 +296,25 @@ dbs_info_callback({error, Reason}, #vacc{resp = Resp0} = Acc) ->
 
 handle_task_status_req(#httpd{method='GET'}=Req) ->
     ok = chttpd:verify_is_server_admin(Req),
-    {Replies, _BadNodes} = gen_server:multi_call(couch_task_status, all),
-    Response = lists:flatmap(fun({Node, Tasks}) ->
-        [{[{node,Node} | Task]} || Task <- Tasks]
-    end, Replies),
-    send_json(Req, lists:sort(Response));
+    Jobs = lists:foldl(fun(Type, JobsAcc) ->
```
Comment (garrensmith) [inline src/chttpd/src/chttpd_misc.erl:299]: It would be better to move this into `fabric2_fdb`. `fabric2_fdb:get_active_tasks` or something like that

File: src/couch_views/src/couch_views_indexer.erl [hunk 9]
```diff
@@ -541,3 +559,35 @@ key_size_limit() ->
 
 value_size_limit() ->
     config:get_integer("couch_views", "value_size_limit", ?VALUE_SIZE_LIMIT).
+
+
+active_tasks_info(ChangesDone, DbName, DDocId, LastSeq, DBSeq) ->
+    VS = case LastSeq of
```
Comment (garrensmith) [inline src/couch_views/src/couch_views_indexer.erl:565]: I don't like storing all this active task info in the job. All of this can be generated when CouchDB receive's a `_active_tasks` request. If we keep adding more and more information to the job info means we running the risk of exceeding the FDB value limit. Rather just add the `current_version_stamp` and `db_version_stamp` to the job options. Is `indexer_pid` actually useful? What would a user do with that?

File: src/couch_views/src/couch_views_indexer.erl [hunk 10]
```diff
@@ -483,17 +489,28 @@ report_progress(State, UpdateType) ->
...
         <<"db_name">> := DbName,
         <<"db_uuid">> := DbUUID,
         <<"ddoc_id">> := DDocId,
         <<"sig">> := Sig,
-        <<"retries">> := Retries
+        <<"retries">> := Retries,
+        <<"active_tasks_info">> := ActiveTasks
     } = JobData,
 
+    ChangesDone1 = case maps:get(<<"changes_done">>, ActiveTasks, 0) of
```
Comment (garrensmith) [inline src/couch_views/src/couch_views_indexer.erl:506]: This can be `ChangesDone = maps:get(<<"changes_done">>, ActiveTasks, 0)` The `0` should be the value if it doesn't exist.

File: src/couch_jobs/src/couch_jobs_fdb.erl [hunk 11]
```diff
@@ -355,6 +357,13 @@ get_activity_vs_and_watch(#{jtx := true} = JTx, Type) ->
     end.
 
 
+get_active_jobs(Type) ->
```
Comment (nickva) [inline src/couch_jobs/src/couch_jobs_fdb.erl:360]: Since we are calling it from the outside of couch_jobs let's add this function to couch_jobs module.  There it could look like:

```
get_active_jobs(Tx, Type) ->
    couch_jobs_fdb:tx(couch_jobs_fdb:get_jtx(Tx), fun(JTx) ->
        couch_jobs_fdb:get_active_since(JTx, Type, {versionstamp, 0, 0}})
    end).
```

`Tx` could then be `undefined`, a plain transaction from erldb or a TxDb one from `fabric2_db`

File: src/chttpd/src/chttpd_misc.erl [hunk 12]
```diff
@@ -294,11 +296,25 @@ dbs_info_callback({error, Reason}, #vacc{resp = Resp0} = Acc) ->
...
+        NewJobs = couch_jobs_fdb:get_active_jobs(Type),
+        maps:merge(NewJobs, JobsAcc)
+    end, #{}, ?ACTIVE_TASK_JOB_TYPES),
+    ActiveTasks = maps:map(fun(JobId, _) ->
+        fabric2_fdb:transactional(fun(Tx) ->
```
Comment (nickva) [inline src/chttpd/src/chttpd_misc.erl:304]: We could move most of this to a fabric2, could be in fabric2_active_tasks or similar though we could also make it smaller if say we get the list of active jobs and their data in the same transaction say one per type. Then we know the data is there and don't have to check for `{error, _}`:

```
    maps:map(fun(Type) ->
        couch_jobs_fdb:tx(couch_jobs_fdb:get_jtx(undefined), fun(JTx) ->
            JobIds = couch_jobs:get_active_jobs(JTx, Type),
            lists:foldl(fun(TasksAcc) ->
               {ok, Data} = couch_jobs:get_job_data(JTx, Type, JobId),
               maps:merge(maps:get(<<"active_tasks_info">>, Data, #{}), TasksAcc)
           end, #{}, JobIds)
    end, ?ACTIVE_TASK_JOB_TYPES)

File: src/couch_views/src/couch_views_indexer.erl (hunk 9 above)
Comment (nickva) [reply src/couch_views/src/couch_views_indexer.erl:565]: The reason I had thought of it is that for replication tasks, for example, we we'd have to know to introspect a few other places to know how to add "source", "target", etc. So active tasks would have to start knowing about how to extract fields from various applications. Perhaps we'd want a simple registration like we have for indices in fabric2_index: https://github.com/apache/couchdb/blob/prototype/fdb-layer/src/fabric/src/fabric2_index.erl#L51-L53 - modules (applications) register a callback for _active_tasks handler and call it when building the response? Or, perhaps we can do that as round 2 if we see the values getting close to 100KB size.

File: src/fabric/src/fabric2_fdb.erl [hunk 13]
```diff
@@ -610,6 +612,21 @@ incr_stat(#{} = Db, Section, Key, Increment) when is_integer(Increment) ->
...
+                maps:map(fun(JobId, _) ->
+                    {ok, Data} = couch_jobs:get_job_data(JTx, Type, JobId),
+                    maps:get(<<"active_tasks_info">>, Data, #{})
+                end, JobIds)
+            end),
+        maps:merge(TaskAcc, Tasks)
+    end, #{}, ?ACTIVE_TASK_JOB_TYPES),
+    maps:values(ActiveTasks).
+
+
```
Comment (nickva) [inline src/fabric/src/fabric2_fdb.erl:629]: Wonder if we could make a separate module for active_tasks, since most things in `fabric2_fdb` deal with db handles and we have a few indexing or db expiration modules that have the `fabric2_<...>` pattern already. 

It could be something like `fabric2_active_tasks` and it would encapsulate both the `?ACTIVE_TASK_JOB_TYPES` list, as well as be the only place that knows about <<"active_tasks_info">>.

So it might have a get_active_tasks() function and also another one that update a job's data with an active task status, like say `update_active_task_info(#{} = JobData, #{} = ActiveTaskInfo)` and it would just do `JobData#{<<"active_tasks_info">> => ActiveTaskInfo}`. Another alternative would be to have an `?ACTIVE_TASK_INFO` define in fabric2.hrl. Just a small ergonomic update to make we don't mistype <<"active_tasks_info">> name somewhere...

File: src/couch_jobs/src/couch_jobs_fdb.erl [hunk 14]
```diff
@@ -45,6 +45,7 @@
     get_jtx/0,
     get_jtx/1,
     tx/2,
+    init_jtx/1,
```
Comment (nickva) [inline src/couch_jobs/src/couch_jobs_fdb.erl:48]: Don't think we need this one anymore

File: src/couch_jobs/src/couch_jobs.erl [hunk 15]
```diff
@@ -104,6 +105,12 @@ get_job_state(Tx, Type, JobId) when is_binary(JobId) ->
     end).
 
 
+get_active_jobs(Tx, Type) ->
+    couch_jobs_fdb:tx(couch_jobs_fdb:get_jtx(Tx), fun(JTx) ->
+        couch_jobs_fdb:get_active_since(JTx, Type, {versionstamp, 0, 0})
+    end).
+
+
```
Comment (nickva) [inline src/couch_jobs/src/couch_jobs.erl:113]: Let's make this a separate commit for this if possible

File: src/couch_views/src/couch_views_indexer.erl (hunk 9 above)
Comment (tonysun83) [reply src/couch_views/src/couch_views_indexer.erl:565]: @garrensmith @nickva how about leaving just `changes_done` to the job_data and then computing the rest on the fly? It seems we should be adding an entirely new fabric_active_tasks module which allows the extractions to reside int here

Comment (garrensmith) [reply src/couch_views/src/couch_views_indexer.erl:565]: That sounds good.

File: src/fabric/src/fabric2_fdb.erl (hunk 13 above)
Comment (tonysun83) [reply src/fabric/src/fabric2_fdb.erl]: @nickva created separate modules for active tasks in fabric and couch_views. I got rid of the need to have a `<<"active_tasks_info">>`. Let me know what you think.

File: src/couch_jobs/src/couch_jobs_fdb.erl [hunk 16]
```diff
@@ -33,6 +33,7 @@
 
     get_activity_vs/2,
     get_activity_vs_and_watch/2,
+    get_active_jobs/1,
```
Comment (nickva) [inline src/couch_jobs/src/couch_jobs_fdb.erl:36]: This is not needed anymore? We moves the function to `couch_jobs`

File: src/couch_jobs/src/couch_jobs_fdb.erl [hunk 17]
```diff
@@ -355,6 +356,13 @@ get_activity_vs_and_watch(#{jtx := true} = JTx, Type) ->
     end.
 
 
+get_active_jobs(Type) ->
+    fabric2_fdb:transactional(fun(Tx) ->
+        JTx = init_jtx(Tx),                                                        
+        get_active_since(JTx, Type, {versionstamp, 0, 0})
+    end).
+
+
```
Comment (nickva) [inline src/couch_jobs/src/couch_jobs_fdb.erl:365]: We are calling get_active_since directly from couch_jobs module so we don't need this function anymore.

File: src/fabric/src/fabric2_active_tasks.erl [hunk 18]
```diff
@@ -0,0 +1,45 @@
...
+
+
+get_active_tasks() ->
+    ActiveTasks = lists:foldl(fun(Type, TaskAcc) ->
+        Tasks = couch_jobs_fdb:tx(couch_jobs_fdb:get_jtx(undefined),
+            fun(JTx) ->
+                JobIds = couch_jobs:get_active_jobs(JTx, Type),
+                maps:map(fun(JobId, _) ->
+                    {ok, Data} = couch_jobs:get_job_data(JTx, Type, JobId),
+                    populate_active_tasks(Type, Data)
```
Comment (nickva) [inline src/fabric/src/fabric2_active_tasks.erl:29]: For consistency wonder if we could pass the JobId in as well, so it might be `populate_active_tasks(Type, JobId, JobData)`

File: src/couch_views/src/couch_views_active_tasks.erl [hunk 19]
```diff
@@ -0,0 +1,87 @@
...
+    catch error:database_does_not_exist ->
+        {ok, not_found}
+    end,
+
+    {ok, DDoc} = case fabric2_db:open_doc(Db, DDocId) of
+        {ok, DDoc0} ->
+            {ok, DDoc0};
+        {not_found, _} ->
+            {ok, not_found}
+    end,
```
Comment (nickva) [inline src/couch_views/src/couch_views_active_tasks.erl:47]: A bit worried about this. We could be starting two transactions here for each task item. One for opening the db and other for opening the doc. With a few hundred active tasks and someone say polling `_active_tasks` for monitoring that could add a lot of load to the system just to add `VStamp` and `DbStamp` to the data.

File: src/couch_jobs/src/couch_jobs_fdb.erl (hunk 16 above)
Comment (tonysun83) [reply src/couch_jobs/src/couch_jobs_fdb.erl:36]: ya was rebasing to get the couch_jobs addition as a separate commit and forgot to undo this

**Comment by nickva** [comment]
@tonysun83 looks better! Good improvement. I made a few comments-in line.

One part that I am not sure if it is worth starting two transaction per task item just to compute version stamp and maybe a few other fields. That seems quite expensive if say someone is polling status on a cluster with hundreds of tasks, given that jobs already have the dbs open and all that data available to start with. The reason we decided to do that was to ensure we are below 100KB for JobData, I think?

We'd have to make a convincing case that our active_jobs status section in each JobData would have a chance of pushing it over the 100KB value limit. I tried encoding a large-ish object similar to the one we are adding for the task status:

```
S = #{
    <<"database">> => << <<"x">> || _ <- lists:seq(1,256) >>,
    <<"changes_done">> => 999999999999999999999999999999,
    <<"design_document">> => << <<"d">> || _ <- lists:seq(1,512) >>,
    <<"current_version_stamp">> => <<"1111111111111111111111111111111111111111">>,
    <<"db_version_stamp">> => <<"222222222222222222222222222222222222222222222">>
}.
```

```
>> size(iolist_to_binary(jiffy:encode(S))).
984
```

So we'd be adding at most another 1KB (in the worst case) to get closer to the 100KB limit if we save the status in the job data that doesn't seem that bad perhaps and it would simplify the design quite a bit more.

File: src/couch_views/src/couch_views_active_tasks.erl (hunk 19 above)
Comment (tonysun83) [reply src/couch_views/src/couch_views_active_tasks.erl:47]: Seems we have two issues
1) exceeding the 100 kb limit with JobData
2) too many read requests when active_task processes exist.

The only 3 things we need are really `changes_done`, `view_seq` and `db_seq`. Wonder if adding just those JobData would an okay compromise?

Comment (tonysun83) [reply src/couch_views/src/couch_views_active_tasks.erl:47]: oh, `view_seq` already exists so only really just need to add `db_seq`

**Comment by davisp** [comment]
I'm actually counting three possible transactions between opening the database if its not in cache, one for the reading the design doc, then one for the other two bits of info.

Can someone point me to the discussion where we decided to move away from storing active task info in the job state? I'm not seeing how this is a better approach or why the job state approach is unworkable.

**Comment by tonysun83** [comment]
@davisp : Here was @garrensmith 's concern https://github.com/apache/couchdb/pull/3003#discussion_r457145064. But I think I found a compromise. We really just need to add two new items to the JobData info: <<"changes_done">> and <<"db_seq">>

**Comment by davisp** [comment]
@tonysun83 The history of this PR is super confusing to me. The last update appears to have reverted to storing all of the info back in the job state which means that this new populate_active_tasks is just "format_job_as_active_task". Which is fine I suppose. Though it seems like it'd be a lot more straightforward for the jobs to just store what they want their active_task output to look like in the first place?

Adding a whole registration/callback system for this seems much more complex than the original plan to just spit out the job data as JSON.

**Comment by tonysun83** [comment]
@davisp:  agreed that it's gotten way more complex. I think @nickva thought that the registration callback would help with the sizing issue, and as a side benefit, add flexibility for future `active_task` types (I know replication is next). I'm leaning toward the simpler method like we did previously without the callback but would like  @nickva  to chime in here since he would know what we would need for replication.

**Comment by nickva** [comment]
@tonysun83:  Agree with @davisp, it's not worth having the registration system just for this.

We could still hide the `<<"active_task_info">>` field name with a simple library function like 

```
fabric2_active_tasks:update_active_task_info(JobData, ActiveTaskInfo) ->
     JobData#{?ACTIVE_TASK_INFO => ActiveTaskInfo}.
```

And use the original design.

>  I think @nickva thought that the registration callback would help with the sizing issue

The registration would have helped with encapsulation, if we had a size issue and had to compute the section on the fly to avoid the http handle hard-coding knowledge about the replication and indexing apps.  But I think we determined computing on the fly was not as optimal, and it wasn't adding that much more to the job data size. If we ever do need to compute it on the fly, we could also come back add the complexity later.

File: src/couch_views/src/couch_views_indexer.erl [hunk 20]
```diff
@@ -541,3 +559,27 @@ key_size_limit() ->
...
+
+active_tasks_info(ChangesDone, DbName, DDocId, LastSeq, DBSeq) ->
+    {[
+        {<<"type">>, <<"indexer">>},
+        {<<"database">>, DbName},
+        {<<"changes_done">>, ChangesDone},
+        {<<"design_document">>, DDocId},
+        {<<"current_version_stamp">>, convert_seq_to_stamp(LastSeq)},
+        {<<"db_version_stamp">>, convert_seq_to_stamp(DBSeq)}
+    ]}.
```
Comment (nickva) [inline src/couch_views/src/couch_views_indexer.erl:572]: Minor nit, if you want this could be a map I think.

File: src/couch_jobs/src/couch_jobs.erl [hunk 21]
```diff
@@ -19,6 +19,7 @@
     remove/3,
     get_job_data/3,
     get_job_state/3,
+    get_active_jobs/2,
```
Comment (nickva) [inline src/couch_jobs/src/couch_jobs.erl:22]: Since we are not returning job data, let's call it `get_active_job_ids/2`

File: src/couch_jobs/src/couch_jobs.erl [hunk 22]
```diff
@@ -104,6 +105,12 @@ get_job_state(Tx, Type, JobId) when is_binary(JobId) ->
     end).
 
 
+get_active_jobs(Tx, Type) ->
+    couch_jobs_fdb:tx(couch_jobs_fdb:get_jtx(Tx), fun(JTx) ->
+        couch_jobs_fdb:get_active_since(JTx, Type, {versionstamp, 0, 0})
```
Comment (nickva) [inline src/couch_jobs/src/couch_jobs.erl:110]: If this returns extra info besides JobIds, let's strip it out so it's just a list of JobIds. So it would basically be `[JobId1, JobId2, ...]`. It might make the caller's logic simpler so they don't have to unpack stuff they don't care about.

File: src/fabric/src/fabric2_active_tasks.erl [hunk 23]
```diff
@@ -0,0 +1,28 @@
...
+-export([
+    get_active_tasks/0
+]).
+
+
+-include_lib("couch/include/couch_db.hrl").
+-include("fabric2.hrl").
+
+
+-define(ACTIVE_TASK_JOB_TYPES, [<<"views">>, <<"replication">>]).
```
Comment (nickva) [inline src/fabric/src/fabric2_active_tasks.erl:13]: To avoid hard-coding the types we could just let any job expose the active_task_info object and we'd automatically display it. So in the get_active_tasks we could query all the types first then call get_active_jobs(JTx, Type).

To get all the types we could expose https://github.com/apache/couchdb/blob/prototype/fdb-layer/src/couch_jobs/src/couch_jobs_fdb.erl#L321 in `couch_jobs` as 

```
get_types(Tx) ->
     couch_jobs_fdb:tx(couch_jobs_fdb:get_jtx(Tx), fun(JTx) ->
        couch_jobs_fdb:get_types(JTx)
    end).
```

This fabric app doesn't have to know anything about replication or views and simply reports status for all the jobs which have an `"active_task_info"` object.

File: src/fabric/src/fabric2_active_tasks.erl [hunk 24]
```diff
@@ -0,0 +1,28 @@
+-module(fabric2_active_tasks).
+
+
+-export([
+    get_active_tasks/0
+]).
+
+
+-include_lib("couch/include/couch_db.hrl").
+-include("fabric2.hrl").
```
Comment (nickva) [inline src/fabric/src/fabric2_active_tasks.erl:10]: Are these used?

File: src/couch_views/src/couch_views_indexer.erl [hunk 25]
```diff
@@ -483,17 +489,28 @@ report_progress(State, UpdateType) ->
...
     } = State,
 
     #{
         <<"db_name">> := DbName,
         <<"db_uuid">> := DbUUID,
         <<"ddoc_id">> := DDocId,
         <<"sig">> := Sig,
-        <<"retries">> := Retries
+        <<"retries">> := Retries,
+        <<"active_tasks_info">> := ActiveTasks
```
Comment (nickva) [inline src/couch_views/src/couch_views_indexer.erl:503]: To avoid hard-coding the field name in all the application, and having a chance someone mistyping it in the future, let's have accessor functions in `fabric2_active_tasks` so that's the only place which knows about `<<"active_task_info">>`. Could have something like `fabric2_active_tasks:update_active_task_info(JobData, ActiveTaskInfo) -> JobData1` and `get_active_task_info(JobData) -> ActiveTaskInfo` for example

**Comment by tonysun83** [comment]
@nickva addressed your latest comments and added a test. Will add more tests later and then rebase with a proper commit history.

File: src/fabric/src/fabric2_active_tasks.erl [hunk 26]
```diff
@@ -0,0 +1,35 @@
+-module(fabric2_active_tasks).
```
Comment (tonysun83) [inline src/fabric/src/fabric2_active_tasks.erl:1]: need to add apache license here

File: src/couch_jobs/src/couch_jobs.erl [hunk 27]
```diff
@@ -104,6 +106,23 @@ get_job_state(Tx, Type, JobId) when is_binary(JobId) ->
...
+    couch_jobs_fdb:tx(couch_jobs_fdb:get_jtx(Tx), fun(JTx) ->
+        Since = couch_jobs_fdb:get_active_since(JTx, Type,
+            {versionstamp, 0, 0}),
+        maps:keys(Since)
+    end).
+
+
+-spec get_types(jtx()) -> [job_type()] | {error, any()}.
+get_types(Tx) ->
+     couch_jobs_fdb:tx(couch_jobs_fdb:get_jtx(Tx), fun(JTx) ->
```
Comment (nickva) [inline src/couch_jobs/src/couch_jobs.erl:121]: Minor nit: I think it's a 5 space indent instead of 4

File: src/fabric/src/fabric2_active_tasks.erl [hunk 28]
```diff
@@ -0,0 +1,35 @@
...
+
+-define(ACTIVE_TASK_INFO, <<"active_task_info">>).
+
+
+get_active_tasks() ->
+    couch_jobs_fdb:tx(couch_jobs_fdb:get_jtx(undefined), fun(JTx) ->
+        Types = couch_jobs:get_types(JTx),
+        lists:foldl(fun(Type, TaskAcc) ->
+            JobIds = couch_jobs:get_active_jobs_ids(JTx, Type),
+            Tasks = lists:map(fun(JobId) ->
```
Comment (nickva) [inline src/fabric/src/fabric2_active_tasks.erl:20]: For jobs that don't have an ?ACTIVE_TASK_INFO field, we might want to not return an empty object but just skip it from the result list altogether. Could use a `filtermap/2` maybe and then if TaskInfo is `#{}` return `false`, otherwise return `{true, TaskInfo}` http://erlang.org/doc/man/lists.html#filtermap-2

File: src/fabric/src/fabric2_active_tasks.erl [hunk 29]
```diff
@@ -0,0 +1,35 @@
+-module(fabric2_active_tasks).
+
+
+-export([
+    get_active_tasks/0,
+    get_active_task_info/1,
+
+    update_active_task_info/2
```
Comment (nickva) [inline src/fabric/src/fabric2_active_tasks.erl:8]: Check the file with emilio, it might complain that the order of exports doesn't match the definition order. So might have to put `update_active_task_info/2` first, or change the definition order, whichever looks better.

File: src/couch_views/src/couch_views_indexer.erl [hunk 30]
```diff
@@ -541,3 +562,4 @@ key_size_limit() ->
 
 value_size_limit() ->
     config:get_integer("couch_views", "value_size_limit", ?VALUE_SIZE_LIMIT).
+
```
Comment (nickva) [inline src/couch_views/src/couch_views_indexer.erl:565]: Whitespace-only change, let's skip it if we can.

**Comment by nickva** [comment]
@tonysun83 looks great! awesome work. Just a few minor nits, whitespace and indents and such

File: src/chttpd/src/chttpd_misc.erl [hunk 31]
```diff
@@ -294,11 +294,8 @@ dbs_info_callback({error, Reason}, #vacc{resp = Resp0} = Acc) ->
 
 handle_task_status_req(#httpd{method='GET'}=Req) ->
     ok = chttpd:verify_is_server_admin(Req),
-    {Replies, _BadNodes} = gen_server:multi_call(couch_task_status, all),
-    Response = lists:flatmap(fun({Node, Tasks}) ->
-        [{[{node,Node} | Task]} || Task <- Tasks]
-    end, Replies),
-    send_json(Req, lists:sort(Response));
+    ActiveTasks = fabric2_active_tasks:get_active_tasks(),
```
Comment (garrensmith) [inline src/chttpd/src/chttpd_misc.erl:297]: Awesome. This is great.

File: src/couch_views/src/couch_views_util.erl [hunk 32]
```diff
@@ -276,3 +277,27 @@ is_paginated(#mrargs{page_size = PageSize}) when is_integer(PageSize) ->
 
 is_paginated(_) ->
     false.
+
+
+active_tasks_info(ChangesDone, DbName, DDocId, LastSeq, DBSeq) ->
+    #{
+        <<"type">> => <<"indexer">>,
```
Comment (garrensmith) [inline src/couch_views/src/couch_views_util.erl:284]: I still don't understand why we need this full active task map in jobs? The only values that we don't already store in the jobs data is `DbSeq` and `ChangesDone`. Couldn't you move those two fields into the job_data and then when you call `fabric2_active_tasks:get_active_tasks` create the full map to return to the active task info based off the job data. This map is a duplication of everything in the job data.

Comment (tonysun83) [reply src/couch_views/src/couch_views_util.erl:284]: I think the idea here is that if have different types of tasks, we would need to know what kind of fields in the JobData to extract for each new task. DBSeq and ChangesDone is probably universal, but different tasks might have unique fields and would then require an update to fabric_active_tasks each time.

Ex: Replication has  
   "checkpointed_source_seq": 68585,
   "continuous": false

and probably has other fields we want as part of active tasks, but we'd have to extract it in fabric2_active_tasks also. 

If we keep one standard active_tasks_info across all JobData, then it enforces encapsulation. As Nick showed up, it's only ~1KB.

Comment (garrensmith) [reply src/couch_views/src/couch_views_util.erl:284]: Ok that makes sense. We can maybe revisit that once we have active tasks for replication and see if its work slimming down.

File: src/fabric/src/fabric2_active_tasks.erl [hunk 33]
```diff
@@ -0,0 +1,51 @@
...
+    end).
+
+
+get_active_task_info(JobData) ->
+    #{?ACTIVE_TASK_INFO:= ActiveTaskInfo} = JobData,
+    ActiveTaskInfo.
+
+
+update_active_task_info(JobData, ActiveTaskInfo) ->
+     JobData#{?ACTIVE_TASK_INFO => ActiveTaskInfo}.
```
Comment (nickva) [inline src/fabric/src/fabric2_active_tasks.erl:51]: Tiny nit: 5 indent instead of 4, add a newline at the end of the file as well, just for so GH diff doesn't show the red warning :-P

Comment (tonysun83) [reply src/fabric/src/fabric2_active_tasks.erl:51]: oh, I thought I changed this. weird. thx
